sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/RegistrationKeyManager.java

```diff
@@ -64,84 +94,266 @@ public boolean getLastExecutionSucceeded() {
+            String fromCryptoAlias = CRYPTO_KEY_PREFIX + fromIndex;
+            String fromAuthAlias = AUTH_KEY_PREFIX + fromIndex;
+
+            //This record is cleared or no need to move
+            if (anyEntryMissed(fromCryptoAlias, fromAuthAlias) || toIndex == fromIndex) {
+                continue;
+            }
+            SecretKey cryptoKey = getSecreteKey(fromCryptoAlias);
+            SecretKey authKey = getSecreteKey(fromAuthAlias);
+
+            String toCryptoAlias = CRYPTO_KEY_PREFIX + toIndex;
+            String toAuthAlias = AUTH_KEY_PREFIX + toIndex;
+            long creationTime = getCreationTime(fromCryptoAlias);
+            storingKeyToFiles(cryptoKey, toCryptoAlias, directoryPath, creationTime);
+            storingKeyToFiles(authKey, toAuthAlias, directoryPath, creationTime);
+
+            deleteKeyFromFiles(fromCryptoAlias, directoryPath);
+            deleteKeyFromFiles(fromAuthAlias, directoryPath);
+            toIndex++;
+        }
+    }
+
+    // Checking if any entries missed
+    private boolean anyEntryMissed(String cryptoKeyAlias, String authKeyAlias) {
+        boolean cryptoKeyMissing = getSecreteKey(cryptoKeyAlias) == null;
+        boolean authKeyMissing = getSecreteKey(authKeyAlias) == null;
+        boolean cryptoTimeMissing = getCreationTime(cryptoKeyAlias) == null;
+        boolean authTimeMissing = getCreationTime(authKeyAlias) == null;
+        return cryptoKeyMissing | authKeyMissing | cryptoTimeMissing | authTimeMissing;
     }
 
     //Loads data from key store file to key store entry
-    private void load(String path) {
-        try (FileInputStream fis = new File(path).exists() ? new FileInputStream(path) : null) {
-            registrationKeyStore.load(fis);
+    private void load(String directoryPath) {
+        char[] password = getKeyStorePassword();
+        String keyStorePath = directoryPath + KEY_STORE_POSTFIX;
+        try (FileInputStream fis = new File(keyStorePath).exists() ? new FileInputStream(keyStorePath) : null) {
+            keyStore.load(fis, password);
+        } catch (IOException | CertificateException | NoSuchAlgorithmException e) {
+            throw clientLogger.logExceptionAsError(new RuntimeException("Failed to load key store", e));
+        }
+
+        String keyCreationTimeStorePath = directoryPath + KEY_CREATION_TIME_POSTFIX;
+        try (FileInputStream fis2 = new File(keyCreationTimeStorePath).exists() ? new FileInputStream(keyCreationTimeStorePath) : null) {
+            keyCreationTimeStore.load(fis2);
         } catch (IOException e) {
-            clientLogger.logExceptionAsError(new RuntimeException("Failed to load key store", e));
+            throw clientLogger.logExceptionAsError(new RuntimeException("Failed to load key creation time store", e));
         }
     }
 
-    private String secretKeyToStr(SecretKey secretKey) {
-        return Base64Util.encodeToString(secretKey.getEncoded());
+    private char[] getPassword() {
```

Comment:
What's the difference between this and `getKeyStorePassword()`?

Reply:
Also, same question about using a hard-coded value.

Reply:
This one is a duplicate. I am going to remove it